### PR TITLE
Provide password to volume create command

### PR DESCRIPTION
**Pull request by davidwilliamson**
**- What I did**
Updated docs to show use of password option when creating docker volume using sshfs plugin. 

This is necessary because: depending on the host configuration between the docker engine and the remote host we wish to mount using sshfs, a password may be required.

* Confirmed syntax for optional parameter with @vieux
* Verified change works as desired with engine running on Ubuntu host and mounting a volume from a file system on a Mac.

**- How to verify it**
Follow the revised example

**- Description for the changelog**
Updated docs to show use of password option when creating docker volume using sshfs plugin. 

Signed-off-by: David Williamson <david.williamson@docker.com>

cc: @vieux

### CODE REVIEW

**Comment by vieux** [comment]
LGTM

**Comment by vieux** [comment]
ping @mstanleyjones

**Comment by stevvooe** [comment]
This should probably be an exception, as this wouldn't be the best default configuration. Something like "Here, use this if you need to pass a password".

**Comment by SvenDowideit** [comment]
LGTM :)

**Comment by cpuguy83** [comment]
Also a really bad idea since this password will be stored in plain text.

**Comment by thaJeztah** [comment]
hm, yes, I have to agree with @stevvooe and @cpuguy83 here; setting passwords this way is probably not a good thing to use in the documentation.

**Comment by mstanleyjones** [comment]
I'm -1 on this. Can you figure out a way to use Docker secrets to store the password?

**Comment by mstanleyjones** [comment]
cc/ @cyli

**Comment by cyli** [comment]
@mstanleyjones Spoke with @davidwilliamson IRL - looking at allowing an option to be passed which can read from a file (like from `/run/secrets/...`)

**Comment by davidwilliamson** [comment]
I was trying to update our docs to match the documented usage for this plugin. The usage page for this plugin (https://github.com/vieux/docker-volume-sshfs) shows passing the SSH password as part of the `docker volume create` command; without providing a password, the example given in the docs does not work.

I don't believe it is feasible to inject credentials into the driver via secrets as requested by @mstanleyjones . Per @vieux , drivers have no access to docker functionality like secrets, bind-mounted files or environment variables. 

I agree that providing the credentials as an argument via the `docker volume create` command results in the credentials being visible in plaintext via `docker volume inspect`. I updated this PR to provide a warning about exposed secrets; I believe the only other alternative would be to remove this example from the docs.

**Review by thaJeztah — APPROVED** [review]
alright, I'm ok with this variation

LGTM

@cyli alright with you?

File: docs/extend/index.md
Comment (cyli) [inline docs/extend/index.md]: Maybe "exposed as plaintext when inspecting the volume" here, since "in the volume" sounds to me like a container that attaches to the volume can see the password as a file or something.

File: docs/extend/index.md
Comment (cyli) [inline docs/extend/index.md]: Can do something like `$(cat file_containing_password)` or something, so at least then it wouldn't be in the bash history as well?

**Comment by cyli** [comment]
If we wanted to keep using the sshfs plugin, I agree this seem to actually make it work (unless we want to tell people to run ssh servers with empty passwords permitted to make the example succeed, or tell people to build their own sshfs plugin image with their ssh key built into the image).

It'd probably be better if we can convert this example to use a different plugin, though, that doesn't use passwords.  (I mean in the long term, not in this PR.)  Maybe something like https://github.com/vdemeester/docker-volume-ipfs?  Although I think that may require a little work, and sshfs is probably more familiar to readers than ipfs.

In the meantime, I'm ok with this, modulo the comments above.

**Comment by stevvooe** [comment]
@davidwilliamson I just don't understand why our first suggestion wouldn't be to use ssh keys. Seems like you drop the key in the right location and it will all work without having a password on the main example. This should be a secondary example like "And if in 2017, you are still using passwords for ssh, here's how".

**Comment by davidwilliamson** [comment]
@stevvooe: I'm not sure what location for the user's SSH keys you are referring to. 

Per @vieux, plugins execute from inside a runc container. AFAIK, a volume driver plugin has _no_ ability to do any of the following:
* read a docker secret
* read a file from the host's file system
* access a bind mounted volume at container create or service start
* read an environment variable

Is there some other approach that you have in mind for using this plugin without requiring a password?

**Comment by stevvooe** [comment]
@davidwilliamson Typically, a user may have their ssh keys in `~/.ssh/`. For these to work for this sshfs mount, they would need to be available to whatever user is running that mount. In this case, I _think_ it is the daemon user.

Using a password here seems both wrong and dangerous. Providing that as the primary example is irresponsible.

**Comment by thaJeztah** [comment]
@stevvooe the daemon may be remote, so no key there.

We should look into these use cases though, especially for future swarm plugins; plugins may need to get support for secrets (if possible)

**Comment by cyli** [comment]
I think @davidwilliamson has done so already, but I just wanted to confirm with @vieux that, as the plugin exists currently, there is no way to get at files on the daemon host system.  Is that correct?

If so, then it seems like before we can tell users to use ssh keys, there is a hard requirement on having a discussion on how/whether the sshfs plugin could be changed to support that.  There are drawbacks, as @thaJeztah points out, such as on a remote daemon, and as @davidwilliamson pointed out IRL, that the ssh keys would probably have to be copied or linked to a fixed location (since if we can do mounts in plugins, the [plugin manifest](https://docs.docker.com/engine/extend/manifest/) docs suggest that the source directory is fixed).

And agree with @thaJeztah, having secret support for swarm plugins in the future would nicely get around this problem altogether.

**Comment by davidwilliamson** [comment]
I submitted a PR to @vieux to add support for SSH keys to the `vieux/sshfs` plugin.

`https://github.com/vieux/docker-volume-sshfs/pull/10`

I welcome any comments on that PR.

If/when that PR is merged, I will update this PR to show using the plugin with a user's SSH keys as the preferred approach.

**Comment by stevvooe** [comment]
@cyli This might be pretty easy to support with secrets+templating. If we create a "secret" template directive, that can populate the secret, allowing keys to be distributed into the mounts. This would be limited to services, but might do the trick.

**Comment by LK4D4** [comment]
@cyli would you mind to review docs? Thanks!

**Comment by cyli** [comment]
I have reviewed https://github.com/vieux/docker-volume-sshfs/pull/10 - if we wanted to get that in first, then the docs no longer have to reflect the password.

Otherwise, if we didn't want the plugin to mount anything from the host directory and wanted to wait on swarm plugins, then I'm ok with this change because otherwise the docs wouldn't work, given that the docs also provide a warning.

**Comment by LK4D4** [comment]
@cyli up to you to decide :)

**Comment by davidwilliamson** [comment]
We could not reach a consensus on a proposed change to the volume plugin (https://github.com/vieux/docker-volume-sshfs/pull/10), so that PR has been closed.

I will leave it up to the maintainers to decide whether to merge this PR as is, or close it and leave the docs as they are.

**Comment by LK4D4** [comment]
ping @cyli again, your last word!

**Comment by thaJeztah** [comment]
Was talking with @cyli and I think it was ok to merge this for now, but we should look for an alternative plugin as an example if possible (let me double check with @cyli)

**Comment by cyli** [comment]
@thaJeztah That is how I feel yes :)

**Comment by thaJeztah** [comment]
ok, let's merge this, thanks!

**Comment by cyli** [comment]
Am working on trying to find an alternative plugin now

**Pull request merged**
